components/entitlement/org.wso2.carbon.identity.entitlement/src/main/java/org/wso2/carbon/identity/entitlement/dao/JDBCSubscriberDAOImpl.java

```diff
@@ -0,0 +1,372 @@
+                }
+            }
+        } catch (SQLException e) {
+            throw new EntitlementException("Error while checking subscriber existence", e);
+        }
+    }
+
+    private void insertSubscriber(String subscriberId, PublisherDataHolder holder, PublisherPropertyDTO[] propertyDTOs)
+            throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(true);
+        try (NamedPreparedStatement createSubscriberPrepStmt = new NamedPreparedStatement(connection,
+                CREATE_SUBSCRIBER_SQL);
+             NamedPreparedStatement createSubscriberPropertiesPrepStmt = new NamedPreparedStatement(connection,
+                     CREATE_SUBSCRIBER_PROPERTIES_SQL)) {
+            createSubscriberPrepStmt.setString(ENTITLEMENT_MODULE_NAME, holder.getModuleName());
+            createSubscriberPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+            createSubscriberPrepStmt.setInt(TENANT_ID, tenantId);
+            createSubscriberPrepStmt.executeUpdate();
+
+            for (PublisherPropertyDTO dto : propertyDTOs) {
+                if (dto.getId() != null && StringUtils.isNotBlank(dto.getValue())) {
+
+                    createSubscriberPropertiesPrepStmt.setString(PROPERTY_ID, dto.getId());
+                    createSubscriberPropertiesPrepStmt.setString(DISPLAY_NAME, dto.getDisplayName());
+                    createSubscriberPropertiesPrepStmt.setString(PROPERTY_VALUE, dto.getValue());
+                    createSubscriberPropertiesPrepStmt.setBoolean(IS_REQUIRED, dto.isRequired());
+                    createSubscriberPropertiesPrepStmt.setInt(DISPLAY_ORDER, dto.getDisplayOrder());
+                    createSubscriberPropertiesPrepStmt.setBoolean(IS_SECRET, dto.isSecret());
+                    createSubscriberPropertiesPrepStmt.setString(MODULE, dto.getModule());
+                    createSubscriberPropertiesPrepStmt.setString(SUBSCRIBER_ID, subscriberId);
+                    createSubscriberPropertiesPrepStmt.setInt(TENANT_ID, tenantId);
+
+                    createSubscriberPropertiesPrepStmt.addBatch();
+                }
+            }
+            createSubscriberPropertiesPrepStmt.executeBatch();
+            IdentityDatabaseUtil.commitTransaction(connection);
+        } catch (SQLException e) {
+            IdentityDatabaseUtil.rollbackTransaction(connection);
+            throw new EntitlementException("Error while inserting subscriber details", e);
+        } finally {
+            IdentityDatabaseUtil.closeConnection(connection);
+        }
+    }
+
+    private void updateSubscriber(String subscriberId, PublisherDataHolder holder, PublisherPropertyDTO[] propertyDTOs,
+                                  PublisherDataHolder oldHolder) throws EntitlementException {
+
+        int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
```

Comment:
Shall we pass `tenantId` as a param to keep consistency of other pure DAO methods?